src/cache.py

```diff
@@ -0,0 +1,292 @@
+
+    ACCOUNTS = "accounts"
+    PERMISSION_SETS = "permission_sets"
+
+
+def _get_ttl_timestamp(ttl_minutes: int) -> int:
+    """Calculate the TTL timestamp for DynamoDB.
+
+    Args:
+        ttl_minutes: TTL in minutes
+
+    Returns:
+        Unix timestamp when the item should expire
+    """
+    return int(time.time()) + (ttl_minutes * 60)
+
+
+def _is_cache_valid(item: dict[str, Any]) -> bool:
+    """Check if a cached item is still valid.
+
+    Args:
+        item: DynamoDB item
+
+    Returns:
+        True if the cache is valid, False otherwise
+    """
+    if "ttl" not in item:
+        return False
+
+    ttl = int(item["ttl"])
+    current_time = int(time.time())
+    return current_time < ttl
+
+
+def get_cached_accounts(
+    dynamodb_client: DynamoDBClient,
+    cache_config: CacheConfig,
+) -> Optional[list[Account]]:
+    """Get cached accounts from DynamoDB.
+
+    Args:
+        dynamodb_client: DynamoDB client
+        cache_config: Cache configuration
+
+    Returns:
+        List of cached accounts or None if cache miss or invalid
+    """
+    if not cache_config.enabled:
+        logger.debug("Cache is disabled, skipping cache lookup")
+        return None
+
+    try:
+        response = dynamodb_client.query(
+            TableName=cache_config.table_name,
+            KeyConditionExpression="cache_key = :cache_key",
+            ExpressionAttributeValues={
+                ":cache_key": {"S": CacheKey.ACCOUNTS},
+            },
+        )
+
+        items = response.get("Items", [])
+        if not items:
+            logger.info("Cache miss for accounts")
+            return None
+
+        # Combine all account items (there might be multiple items due to pagination)
+        all_accounts = []
+        for item in items:
+            if not _is_cache_valid(item):
+                logger.info("Cache expired for accounts")
+                return None
+
+            accounts_data = json.loads(item["data"]["S"])
+            all_accounts.extend([Account.parse_obj(acc) for acc in accounts_data])
+
+        logger.info(f"Cache hit for accounts, found {len(all_accounts)} accounts")
+        return all_accounts
+
+    except Exception as e:
+        logger.warning(f"Failed to get cached accounts: {e}", extra={"error": str(e)})
+        return None
+
+
+def set_cached_accounts(
+    dynamodb_client: DynamoDBClient,
+    cache_config: CacheConfig,
+    accounts: list[Account],
+) -> None:
+    """Store accounts in DynamoDB cache.
+
+    Args:
+        dynamodb_client: DynamoDB client
+        cache_config: Cache configuration
+        accounts: List of accounts to cache
+    """
+    if not cache_config.enabled:
+        logger.debug("Cache is disabled, skipping cache write")
+        return
+
+    try:
+        ttl = _get_ttl_timestamp(cache_config.ttl_minutes)
+        accounts_data = [account.dict() for account in accounts]
+
+        # Store all accounts in a single item
+        dynamodb_client.put_item(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential NoSQL Injection detected. Untrusted input is being used in a NoSQL database
operation without proper sanitization. NoSQL databases like MongoDB and DynamoDB are not
immune to injection attacks. User input should never be directly interpolated into query
objects or strings. Instead, use parameterized queries, proper sanitization techniques,
or type conversion specific to the NoSQL database you&#39;re using. For MongoDB, consider using
ObjectId() for IDs and validated operators. For DynamoDB, use `boto3.dynamodb.conditions`
classes for safe query construction. Learn more - https://cwe.mitre.org/data/definitions/943.html